cache/cache.go

```diff
@@ -0,0 +1,160 @@
+// Copyright 2015 Canonical Ltd.
+// Licensed under the LGPLv3, see LICENCE file for details.
+
+// Package cache provides a simple caching mechanism
+// that limits the age of cache entries and tries to avoid large
+// repopulation events by staggering refresh times.
+package cache
+
+import (
+	"math/rand"
+	"sync"
+	"time"
+
+	"gopkg.in/errgo.v1"
+)
+
+type entry struct {
+	value  interface{}
+	expire time.Time
+}
+
+// Key represents a cache key. It must be a comparable type.
+type Key interface{}
+
+// Cache holds a time-limited cache of values for arbitrary keys.
+type Cache struct {
+	maxAge time.Duration
+
+	// mu guards the fields below it.
+	mu sync.Mutex
+
+	// expire holds when the cache is due to expire.
+	expire time.Time
+	// We hold two maps so that can avoid scanning through all the
+	// items in the cache when the cache needs to be refreshed.
+	// Instead, we move items from old to new when they're accessed
+	// and throw away the old map at refresh time.
+	old, new map[Key]entry
+}
+
+// New returns a new Cache that will cache items for
+// at most maxAge.
```

Comment:
There are no lower bounds. It's rounded up to 2ns but that should affect the caller in practice.